vast/opensetAlgos/WPL.py

```diff
@@ -0,0 +1,140 @@
+    )
+    WPL_params.add_argument(
+        "--distance_multiplier",
+        nargs="+",
+        type=float,
+        default=[1.0],
+        help="distance multiplier to use default: %(default)s",
+    )
+    WPL_params.add_argument(
+        "--default_scale",
+        nargs="+",
+        type=float,
+        default=[1.0],
+        help="Weibull scale to use when the number of uniqe element is less than 5 default: %(default)s",
+    )
+    WPL_params.add_argument(
+        "--default_shape",
+        nargs="+",
+        type=float,
+        default=[0.1],
+        help="Weibull shape to use when the number of uniqe element is less than 5 default:  %(default)s",
+    )
+    return parser, dict(
+        group_parser=WPL_params,
+        param_names=("tailsize", "distance_multiplier", "default_scale", "default_shape"),
+        param_id_string="TS_{}_DM_{:.4f}_SC_{:.4f}_SH_{:.4f}",
+    )
+
+
+def fit_high(distances, distance_multiplier, tailsize, default_shape, default_scale):
+    distances = torch.unique(distances)
+    if tailsize <= 1:
+        tailsize = min(tailsize * distances.shape[1], distances.shape[1])
+    tailsize = int(min(tailsize, distances.shape[1]))
+    mr = weibull.weibull()
+    if distances.shape[1] < 5:
+        pass
+    else:
+        mr.FitHigh(distances.double() * distance_multiplier, tailsize, isSorted=False)
+    mr.tocpu()
+    return mr
+
+  
+def WPL_Training(
+    pos_classes_to_process: List[str],
+    features_all_classes: OrderedDict[str, torch.Tensor],
+    args,
+    gpu: int,
+    models=None,
+) -> Iterator[Tuple[str, Tuple[str, dict]]]:
+    """
+    :param pos_classes_to_process: List of class names to be processed by this function in the current process class.
+    :param features_all_classes: features of all classes, note the classes in pos_classes_to_process can be a subset of the keys for this ordered dictionary
+    :param args: A named tuple or an argument parser object containing the arguments mentioned in the WPL_Params function above.
+    :param gpu: An integer corresponding to the gpu number to use by the current process.
+    :param models: Not used during training, input ignored.
+    :return: Iterator(Tuple(parameter combination identifier, Tuple(class name, its evm model)))
+    """
+    with torch.no_grad():
+        for pos_cls_name in pos_classes_to_process:
+            features = features_all_classes[pos_cls_name].clone().to(f"cuda:{gpu}")
+            assert args.dimension == features.shape[1]
+
+            center = torch.mean(features, dim=0).to(f"cuda:{gpu}")
+            distances = torch.abs(features - center.view(1,args.dimension).repeat(features.shape[0], 1))
+            for tailsize, distance_multiplier, default_shape, default_scale  in itertools.product(
+                args.tailsize, args.distance_multiplier, args.default_shape, args.default_scale
+            ):
+                  weibull_list = list()
+                  for k in args.dimension:
```

Comment:
You should also be able to use it to return multiple weibulls at once. If `distances` is a `nxk` tensor, where in your case `n` is the number of dimensions, then it can straightaway provide you `n` weibulls in a single mr object which will reduce both training and inference time.